File: other_algorithms/limit/limit.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <sstream>
#include <optional>

using namespace std;

optional<int> parse_input(const string& input) {
    // Pressing the Enter key without entering a value
    if (input.empty()) {
        return nullopt;
    }

    stringstream ss(input);
    int number;
    ss >> number;

    // insert invalid value
    if (ss.fail() || !ss.eof()) {
        return nullopt; 
    }

    return number;
}

tuple<int, int> check_minmax(const vector<int>& array, optional<int> min_val_opt, optional<int> max_val_opt) {
    int actual_min = *min_element(array.begin(), array.end());
    int actual_max = *max_element(array.begin(), array.end());
    
    int min_val = min_val_opt.value_or(actual_min);
    int max_val = max_val_opt.value_or(actual_max);

    if (min_val > max_val) {
        swap(min_val, max_val);
    }

    return {min_val, max_val};
}


int main(int argc, char** argv){
    std::string input_str;
    vector<int> array;
    vector<int> result;

    
    std::cout << "Enter the array elements (separate elements with commas): ";
    getline(std::cin, input_str);
    stringstream ss(input_str);
    std::string temp;
    char delimiter = ',';
    while (getline(ss,temp, delimiter))
    {
        array.push_back(std::stoi(temp));
    }

    if(array.empty()){
        std::cerr << "Array is empty." << std::endl;
        return 0;
    }

    optional<int> min_val, max_val;

    std::cout << "Enter Min Value: ";
    getline(std::cin, input_str);
    min_val = parse_input(input_str);

    std::cout << "Enter Max Value: ";
    getline(std::cin, input_str);
    max_val = parse_input(input_str);

    auto [filtered_min, filtered_max] = check_minmax(array, min_val, max_val);
    std::cout << "Min Value: " << filtered_min << " - " << "Max Value: " << filtered_max << std::endl;    

    for(int item: array){
        if(item >= filtered_min && item <= filtered_max)
            result.push_back(item);
    }

    for(int i: result)
        cout << i << " ";
    
    return 0;
}

File: other_algorithms/limit/limit.py
from typing import List


def limit_from_scratch(arr:List, min_val:int=None, max_val:int=None):
    if len(arr) == 0:
        return None
    if min_val is None:
        min_val = min(arr)
    if max_val is None:
        max_val = max(arr)
    result = []
    for items in arr:
        if items >= min_val and items <= max_val:
            result.append(items)    
    return result     


def simple_limit(arr:List, min_val:int=None, max_val:int=None):
    if len(arr) == 0:
        return None
    if min_val is None:
        min_val = min(arr)
    if max_val is None:
        max_val = max(arr) 
    return list(filter(lambda x: (min_val <= x <= max_val), arr))


def advance_limit(arr:List, min_val:int=None, max_val:int=None):
    min_check = lambda x: True if min_val is None else (x >= min_val)
    max_check = lambda x: True if max_val is None else (x <= max_val)
    
    return [val for val in arr if min_check(val) and max_check(val)]
 

array = list(map(int, input("Enter elemnts array: ").split(",")))

min_val = input("Enter Min Value: ")
max_val = input("Enter Max Value: ")

if min_val == "" and max_val == "":
    min_val, max_val = None, None
elif min_val != "" and max_val == "":
    min_val = int(min_val)
    max_val = None
elif min_val == "" and max_val != "":
    min_val = None
    max_val = int(max_val)    
else:
    min_val = int(min_val)
    max_val = int(max_val)

print(limit_from_scratch(array, min_val, max_val))
print(simple_limit(array, min_val, max_val))
print(advance_limit(array, min_val, max_val))
